base/dict.jl

```diff
@@ -258,12 +267,13 @@ function isequal(l::Associative, r::Associative)
 end
 
 function ==(l::Associative, r::Associative)
+    l === r && return true
     if isa(l,ObjectIdDict) != isa(r,ObjectIdDict)
         return false
     end
     if length(l) != length(r) return false end
-    for (key, value) in l
-        if value != get(r, key, secret_table_token)
+    for pair in l
+        if !in(pair, r, ==)
```

Comment:
the type constraint is due to the expectation that an Associative only contains Pairs (e.g. see the error message you get if you look for something else: https://github.com/JuliaLang/julia/pull/13825/files#diff-d4736682004c1996e5d9b1c5c33c8c8eR28)